### PR TITLE
only one hpa can scaler same target

### PR DESCRIPTION
**Pull request by kadisi**
Signed-off-by: kadisi <iamkadisi@163.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
If there are multiple hpa referance same target, only the newly hpa is valid
```

### CODE REVIEW

File: pkg/controller/podautoscaler/horizontal.go [hunk 1]
```diff
@@ -370,6 +370,27 @@ func (a *HorizontalController) reconcileKey(key string) error {
...
+// return false, newlyHpa, nil, if hpav1 is not newly created for same ScaleTargetRef
+// newlyHpa is a hpa object which created after hpav1 and ref same ScaleTargetRef
+func (a *HorizontalController) isLasterHpa(hpav1 *autoscalingv1.HorizontalPodAutoscaler) (bool, *autoscalingv1.HorizontalPodAutoscaler, error) {
+	allHpas, err := a.hpaLister.HorizontalPodAutoscalers(hpav1.Namespace).List(labels.Everything())
+	if err != nil {
+		return false, nil, fmt.Errorf("failed to get all hpas in namespaces %v", hpav1.Namespace)
+	}
+
+	for _, h := range allHpas {
+		if h.Spec.ScaleTargetRef.Kind == hpav1.Spec.ScaleTargetRef.Kind &&
```
Comment (yastij) [inline pkg/controller/podautoscaler/horizontal.go]: I'd refactor this using an affectation `h.Spec.ScaleTargetRef` and `hpav1.Spec.ScaleTargetRef`

Comment (yastij) [inline pkg/controller/podautoscaler/horizontal.go]: I'd refactor this using an affectation for `h.Spec.ScaleTargetRef` and `hpav1.Spec.ScaleTargetRef`

File: pkg/controller/podautoscaler/horizontal.go [hunk 2]
```diff
@@ -494,6 +519,14 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 
 	}
 
+	if !isLaster {
+		desiredReplicas = 0
+		rescale = false
+		setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionFalse, "ScalingDisabled", "scaling is disabled since find new version hpa")
+		msg := fmt.Sprintf("hpa %v is invalid, because we find new hpa %v references same targe %v %v", hpav1.Name, newlyHpa.Name, hpav1.Spec.ScaleTargetRef.Kind, hpav1.Spec.ScaleTargetRef.Name)
```
Comment (yastij) [inline pkg/controller/podautoscaler/horizontal.go]: Use past for event messages.

File: pkg/controller/podautoscaler/horizontal.go (hunk 1 above)
Comment (kadisi) [reply pkg/controller/podautoscaler/horizontal.go]: I'm verry sorry I don't understand what's you mean.

**Comment by DirectXMan12** [comment]
/close

This would need serious discussion to even consider.  I suggest you bring it up on the sig-autoscaling mailing list, in an issue, or in the weekly meeting.

However, this is not likely to land.  We generally aren't in the habit of adding this kind of check in Kubernetes.  Furthermore, this particular implementation is a bit opaque to the user, and exposes creation-time-based ordering into Kubernetes, which we generally reject in Kubernetes.

On top of that, I'm not a huge fan of making the HPA reconcile loop O(n^2).

**Pull request closed without merging**